src/lightning_app/core/work.py

```diff
@@ -457,8 +482,35 @@ def set_state(self, provided_state):
             if isinstance(v, Dict):
                 v = _maybe_create_drive(self.name, v)
             setattr(self, k, v)
+
         self._changes = provided_state["changes"]
-        self._calls.update(provided_state["calls"])
+
+        # Note, this is handled by the flow only.
+        if _is_flow_context():
+            self._cleanup_calls(provided_state["calls"])
+
+        self._calls = provided_state["calls"]
+
+    @staticmethod
+    def _cleanup_calls(calls: Dict[str, Any]):
+        # 1: Collect all the in_progress call hashes
+        in_progress_call_hash = [
+            k for k in list(calls) if k not in (CacheCallsKeys.LATEST_CALL_HASH, CacheCallsKeys.SUCCEEDED_CALL_HASHES)
+        ]
+
+        for call_hash in in_progress_call_hash:
+            # 2: Filter the statuses by timestamp
+            statuses = sorted(calls[call_hash]["statuses"], key=lambda x: x["timestamp"])
+
+            # If the latest status is succeeded, let's drop entirely
+            # the statuses collection associated to this call_hash
+            # and store the call_hash inside the `SUCCEEDED_CALL_HASHES` set.
+            # TODO: Store the statuses within a database in the future
+            if statuses[-1]["stage"] == WorkStageStatus.SUCCEEDED:
+                calls.pop(call_hash)
+                calls[CacheCallsKeys.SUCCEEDED_CALL_HASHES].add(call_hash)
```

Comment:
@tchaton is this `SUCCEEDED_CALL_HASHES` getting cleaned ?

Reply:
No, this is the cache call hashes.